perf: segment Source content into chunked arrays

rjharmon opened the pull request:
- retains reasonable performance when used for tokenizing large content > 10k bytes

rjharmon committed:
perf: segment Source content into chunked arrays

 - retains reasonable performance when used for tokenizing large content > 10k bytes

christianschmitz commented on src/tokens/Source.js:
```diff
@@ -11,9 +11,9 @@
+     * @type {string[][]}
      */
-    content
+    contentIndex
```
`contentIndex` is a slighly confusing name. Could we rename this to e.g. `segmentedContent` ?

rjharmon replied:
made it `contentChunks`

christianschmitz commented on src/tokens/Source.js:
```diff
@@ -26,26 +26,57 @@ export class Source {
      */
-    get length() {
-        return this.content.length
+    length
```
I think all properties should above the constructor, and maybe it makes sense to add the `@readonly` annotation to this property

Sadly I don't know of any formatter that moves properties into the class header automatically

rjharmon replied:
yup, done

christianschmitz commented on src/tokens/Source.js:
```diff
@@ -26,26 +26,57 @@ export class Source {
+     * @param {string[]} content
+     * @returns {string[][]}
+     */
+    segmentContent(content) {
```
I think this utility function shouldn't be a member of this class, but rather be moved into the codec-utils repo, with the segmentSize being a second argument

Also I think a for-loop conveys the intent better:
```ts
const n = content.length
/**
 * @type {string[][]}
 */
const segments = []
    for (let i = 0; i < content.length; i += segmentSize) {
        segments.push(content.slice(i, i + segmentSize));
    }
    return segments;
```

Update: I've added `segmentArray()` to `@helios-lang/codec-utils` v0.1.35, so we can import that here

rjharmon replied:
ok, thanks.  Using that now.

christianschmitz commented on src/tokens/Source.js:
```diff
@@ -26,26 +26,57 @@ export class Source {
+        // one-step split to utf-8 runes in the content
+        const codePoints = [...content]
+        // heuristic for segment size
+        this.segmentSize = Math.max(
```
For consistency I think all new properties here should get their own block of declarations in the class header above the constructor

rjharmon replied:
ah, yes - I had thought of that, but then I forgot to move things around before pushing.

thanks!

christianschmitz commented on src/tokens/Source.js:
```diff
@@ -26,26 +26,57 @@ export class Source {
      */
     constructor(content, options = {}) {
-        this.content = content
+        this.rawContent = content
```
Could you change the name of this property back to `content`? I think I have some downstream packages depend on that

rjharmon replied:
done

christianschmitz commented on src/tokens/Source.js:
```diff
@@ -76,34 +107,45 @@ export class Source {
+    /**
+     * @type{[number, number][] | undefined}
+     */
+    lineLengths
```
I would change the name of this property to `_lineLengths` and make it `@private`, then change its type to `Option<[number, number][]>` and initialize it explicitly in the constructor using `None` imported from @helios-lang/type-utils.

The `calcLineLengths()` function I would change to `get lineLengths()`, which returns either the cached result from the private property, or does the calculation (assigning to the private property in the process).

That way any function that calls `this.lineLengths` doesn't have to worry about the caching.

rjharmon replied:
They're actually line-ending locations; renamed.

The first number in the tuple wasn't actually used/needed, so I changed the type to `number[]` too.

christianschmitz commented on src/tokens/Source.js:
```diff
@@ -76,34 +107,45 @@ export class Source {
+     * Calculates the length of each line and the total length up to that line
+     * @returns {[number, number][]}
+     */
+    calcLineLengths() {
```
See comment about `lineLenghts`

rjharmon replied:
thanks for this

christianschmitz commented on src/tokens/Source.js:
```diff
@@ -76,34 +107,45 @@ export class Source {
-            } else {
-                col += 1
-            }
+        const lengths = this.lineLengths || this.calcLineLengths()
```
See comment above about `lineLengths`

rjharmon replied:
done

christianschmitz reviewed (changes requested):
Thanks for contributing this, I've requested a few changes in order to maintain consistency with the rest of the code-base

rjharmon committed:
perf improvement code refinements

 - use imported utility for segmenting content to chunks
   ... with updated dependency
 * moved declarations before constructor
 * removed excess variable whose name was awkward
 * adjusted some property names
 * changed line-ending-locations[][_,x:number] to just a `number[]`

rjharmon commented:
ok I believe all should be in good shape now.

rjharmon commented:
ok I believe all should be in good shape now.

christianschmitz commented:
Thanks for making the requested changes
